enhancements/network/service-egress-traffic-steering.md

```diff
@@ -26,7 +26,8 @@ tracking-link:
 Some external systems that communicate with applications running on the Kubernetes cluster through a LoadBalancer service expect that the source IP of egress traffic originating from the pods backing the service is identical to the destination IP they use to reach them - i.e the LoadBalancer's ingress IP.
 This behavior requires that service access from outside the cluster will only work to a designated node. This is contradictory to how default services work where a service may be accessed via any node.
 
-By annotating a LoadBalancer service, users could request that the source IP of egress packets originating from all of the pods that are endpoints of it would be its ingress IP.
+By introducing a new CRD `EgressService`, users could request that the source IP of egress packets originating from all of the pods that are endpoints of a LoadBalancer service would be its ingress IP.
+The CRD will be Namespaced, each named as the LoadBalancer Service that should be modified.
```

Comment:
Do we really want to use name here to do this? I guess the advantage is you make sure each service only has a single EgressService.

But what if you want 1 EgressService to be used for multiple services. Seems like maybe a labelSelector would be better to match on services? Then in ovnk we only allow 1 EgressService per service?

Reply:
My idea was that using a 1-1 mapping makes it clear which service acts as an "egress service" (+ that section where I wrote the considerations)

if we think in the future that it is beneficial to have a 1-many relation what I have in mind is maybe creating a new resource ("EgressServiceProfile / whatever") that uses selectors which results in ovn-k creating the relevant egress services objects (similar to a deployment-pod relation)

Reply:
I'm not strongly against using the proposal of name mapping, but it does seem awkward to me.

@msherif1234 @jcaamano wdyt?

Reply:
I would not match using the CRD name itself. I would match on something that is in the CRD spec, regardless of it being 1 on 1 or 1 to many. In the spec we can have a service name or a service selector, whatever we need for now. And it will be extensible in the future without needing a new CRD (much like the pod has nodeName, nodeSelector, ...).

Reply:
I agree that it might seem awkward/odd, but the reason I think it makes sense here (In addition to what I wrote in the "Design Details" section) is that the CRD could have been replaced by a field in the `Service` spec and when that's not an option it seems that having a name mapping resembles it (and the old annotation) the most.

It provides a clear/visible way to understand which service is configured and prevents ambiguous configurations.
As a sidenote, I think it is cleaner to implement and reduces the amount of corner-cases (might be more performant as well because it allows us to always process one EgressService per service).

As for having serviceSelector/serviceName in the spec instead, I think we should focus on a name for now because selectors requires more care such as changing the status field to accommodate multiple service-node relations.

I'm open to changing this, but would like to see if I'm missing something crucial here as imo it seems like a good option. wdyt @jcaamano @trozet?

Reply:
So I asked @danwinship for his opinion and he rightly pointed out that services -> endpoints is a name mapping, so for now I guess this seems appropriate to keep things simple. I'll approve the enhancement and wait on @jcaamano to give his thoughts/decision on final approval.

Reply:
OK, I don't have a strong opinion either against the name mapping, I guess we can always add a `serviceName` to the CRD spec in the future and stay backward compatible. Although I don't see how having a serviceName in the spec would complicate things further now.

But for me clearly using name mapping is not the best API design. A resource name is for the user to use as he pleases for administrative purposes and giving to it more meaning than that is not usually the best idea from an API standpoint. But using name mapping might be a well established thing in kubernetes?

Anyway, let's go with what we have now :+1: